File: src/schema/query.ts
import { z } from "zod"

export const DateRangeSchema = z.object({
  date_from: z.string().describe("The start date of the date range. Could be a date string or a relative date string like '-7d'"),
  date_to: z.string().describe("The end date of the date range. Could be a date string or a relative date string like '-1d'"),
})

export const HogQLFiltersSchema = z.object({
  dateRange: DateRangeSchema.optional(),
})

export const HogQLQuerySchema = z.object({
  kind: z.literal("HogQLQuery"),
  query: z.string(),
  explain: z.boolean().optional(),
  filters: HogQLFiltersSchema.optional(),
})

export const InsightQuerySchema = z.object({
  kind: z.literal("DataVisualizationNode"),
  source: HogQLQuerySchema,
})

export type InsightQuery = z.infer<typeof InsightQuerySchema>;
export type HogQLQuery = z.infer<typeof HogQLQuerySchema>;
export type HogQLFilters = z.infer<typeof HogQLFiltersSchema>;
export type DateRange = z.infer<typeof DateRangeSchema>;

File: python/schema/query.py
from pydantic import BaseModel, Field


class DateRange(BaseModel):
    date_from: str = Field(description="The start date of the date range. Could be a date string or a relative date string like '-7d'")
    date_to: str = Field(description="The end date of the date range. Could be a date string or a relative date string like '-1d'")


class HogQLFilters(BaseModel):
    dateRange: DateRange | None = None


class HogQLQuery(BaseModel):
    kind: str = "HogQLQuery"
    query: str
    explain: bool | None = None
    filters: HogQLFilters | None = None


class InsightQuery(BaseModel):
    kind: str = "DataVisualizationNode"
    source: HogQLQuery
